### PR TITLE
FLAME on SHARC (libmboard, xparser, GSL)

### PR DESCRIPTION
**Pull request by willfurnass**
Documentation, install scripts, install logs and modulefiles.  As requested by @md1ashu.  Awaiting review.

### CODE REVIEW

File: sharc/software/apps/flame.rst [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+#. Run the ``xparser`` utility on the XMML (*X Machine Markup Language*) XML file corresponding to your model 
+   (which references the C source code files containing your state transition functions): ::
+
+      xparser -p model.xml
+
+   This generates code that can be compiled to provide a FLAME simulation program.
+   Here we passed the ``-p`` argument to ``xparser``, which tells it to generate simulation source code that can be run in parallel using MPI.
+   Alternatively you can specify ``-s`` if you want a serial simulation program instead.
+   In addition, you can specify a ``-f`` flag if you want to generate an optimise (*final*) build of your FLAME simulation program.
```
Comment (dmrogers75) [inline sharc/software/apps/flame.rst]: Change "optimise" to "optimised".

File: sharc/software/apps/flame.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+   
+      mpirun ./main 5000 its/0.xml
+
+   This will distribute work over all the CPU cores you requested for your job.
+
+
+Example using MPI
+-----------------
+
+#. Download and unpack the some example FLAME models using: ::
```
Comment (dmrogers75) [inline sharc/software/apps/flame.rst]: Place "some" in quotes or backticks?

File: sharc/software/apps/flame.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+#. Create a subdirectory for (XML) state information: ::
+
+      mkdir its
+      cp 0.xml its/
+
+#. Create a batch job submission script in the current working directory, ``model_05``, called ``flame_example_model_05.sge`` containing: ::
+   
+      #!/bin/bash
+      #$ -pe mpi 4
```
Comment (dmrogers75) [inline sharc/software/apps/flame.rst]: Runtime and memory per core required for FLAME? Or are defaults OK?

File: sharc/software/install_scripts/apps/xparser/0.17.1/gcc-6.2/install.sh [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+tar -zxf ${tarball_name}
+pushd ${tarball_top_level_dir}
+
+###################### Configure and install #################################
+
+make clean
+make
+module load libs/libmboard/0.3.1/gcc-6.2-openmpi-2.1.1
+sed -i "s#/Users/stc/workspace/libmboard#${prefix}#" tests/Makefile
+#CPATH=${prefix}/include:${CPATH} LD_LIBRARY_PATH=${prefix}/lib:${LIBRARY_PATH} LIBMBOARD_DIR= make MAKEOVERRIDES=LIBMBOARD_DIR test
```
Comment (dmrogers75) [inline sharc/software/install_scripts/apps/xparser/0.17.1/gcc-6.2/install.sh]: Are these a note of env vars declared somewhere and the make command?

**Comment by willfurnass** [comment]
@dmrogers75 Thanks for the suggestions.

Just discovered an issue: FLAME simulations that use MPI hang after what appears to be an arbitrary  number of iterations.   Forcing `mpirun` to timeout after a certain interval (`--get-stack-traces --timeout 600 --verbose`) shows that  processes are waiting on the message board to sync (`MB_SyncComplete`) with respect to various agent types.  

Haven't yet been able to repeat this using serial FLAME simulations.

**Comment by desmfryan** [comment]
Hi Will

I have tested your Flame install on ShARC and it run's fine with one of the test models (from Aban I think) I have.

However, my first attempt failed since I only module loaded the xparser with no knowledge that the libmboard library needed to be installed (parallel or serial) separately. I only realised this when I did a module load xparser in an interactive session (then you see your message). 

My suggestion would be load the parallel libmboard with the xparser by default, and then give instructions in the documentation of what to do if serial execution is required.

Cheers,

Des

**Comment by desmfryan** [comment]
@willfurnass see above comments...thanks.

File: sharc/software/apps/flame.rst (hunk 3 above)
Comment (willfurnass) [reply sharc/software/apps/flame.rst]: Defaults fine.

**Comment by willfurnass** [comment]
* Info on the need to load both the xparser and libmboard is in the docs
* @mondus tells me that the MPI error is likely an application bug rather than a install/configuration issue.

**Pull request merged**